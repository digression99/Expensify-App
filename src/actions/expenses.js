import uuid from 'uuid';
import database from '../firebase/firebase';

// export const addExpense = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
//     type : 'ADD_EXPENSE',
//     expense : {
//         id : uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });

export const addExpense = expense => ({
    type : 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => dispatch => {
    const {
        description = '', note = '', amount = 0, createdAt = 0 // default data
    } = expenseData; // destructure.
    const expense = {
        description, note, amount, createdAt
    };
    return database.ref('expenses').push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id : ref.key,
                ...expense
            }));
        });
};

export const removeExpense = ({ id } = {}) => ({
    type : 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id} = {}) => dispatch => {
    // console.log('in start remove expense, id : ', id);
    return database.ref(`expenses/${id}`).remove()
        .then(() => {
            // console.log('in start remove expense, before dispatch, id : ', id);
            dispatch(removeExpense({id}));
        });
};

export const editExpense = (id, updates) => ({
    type : 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => dispatch =>
    database.ref(`expenses/${id}`)
        .update({
            ...updates
        })
        .then(() => {
            dispatch(editExpense(id, updates));
        });

// create startEditExpense (same call signature as editExpense)
// test startEditExpense with 'should edit expenses from firebase'
// use startEditExpense in EditExpensePage instead of editExpense
// adjust EditExpensePage tests

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type : 'SET_EXPENSES',
    expenses
});

// export const startSetExpenses;
// fetch all expenses from database once.
// parse that data into an array.
// dispatch SET_EXPENSES

export const startSetExpenses = (expenses = []) => dispatch => {
    return database.ref('expenses').once('value')
        .then(snapshot => {
            snapshot.forEach(childSnapshot => {
                expenses.push({
                    ...childSnapshot.val(),
                    id : childSnapshot.key
                });
            });
            dispatch(setExpenses(expenses));
        })
};

// create startRemoveExpense (same call signature as removeExpense)
// test startRemoveExpense with 'should remove expenses from firebase' // test if it's null.
// use startremoveExpense in EditExpensePage instread of removeExpense
// adjust EditExpensePage tests


// const printExpenses = snapshot => {
//     const expenses = [];
//     snapshot.forEach(childSnapshot => {
//         expenses.push({
//             ...childSnapshot.val(),
//             id : childSnapshot.key
//         })
//     });
//     console.log(expenses);
// };
//
// database.ref('expenses')
//     .once('value')
//     .then(printExpenses);