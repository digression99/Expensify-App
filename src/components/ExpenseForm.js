import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description : props.expense ? props.expense.description : '',
            note : props.expense ? props.expense.note : '',
            amount : props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt : props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused : false,
            error : ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) { // prevent clearning the date.
            this.setState(() => ({ createdAt }))
        }
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused : focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({error : 'PLEASE, provide description and amount.'}))
        } else {
            this.setState(() => ({error : ''}));
            const {description, amount, note, createdAt} = this.state;

            this.props.onSubmit({
                description,
                amount : parseFloat(amount, 10) * 100,
                createdAt : createdAt.valueOf(),
                note
            });
            // console.log('submitted!');
        }
    };

    render() {
        return (
                <form className="form" action="" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <input type="text"
                           className="text-input"
                           placeholder="Description"
                           autoFocus
                           value={this.state.description}
                           onChange={this.onDescriptionChange}
                    />
                    <input type="number"
                           className="text-input"
                           placeholder="Amount"
                           value={this.state.amount}
                           onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.calendarFocused} // PropTypes.bool
                        onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        name=""
                        id=""
                        cols="30"
                        rows="10"
                        className="textarea"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <div>
                        <button
                            className="button-layout"
                        >
                            {this.props.expense ? "Save Expense" : "Add Expense"}
                        </button>
                    </div>
                </form>
        );
    }
}