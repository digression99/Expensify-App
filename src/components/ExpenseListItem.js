// export a stateless functional component
// render description, amount, createdAt.
// expenselist 에서 이걸 이용해 expenses.map.

import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = //(props) => (
    ({id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            amount : {numeral(amount / 100).format('$0.0.00')},
            createdAt : {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
        {/*actually remove it.*/}
    </div>
);

export default ExpenseListItem;