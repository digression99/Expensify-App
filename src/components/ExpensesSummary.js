import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0.0.00');

    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
        </div>
    );
};

// export const ExpensesSummary = ({expenses}) => (
//     <div>
//         <p>total amount : {numeral(selectExpensesTotal(expenses)).format('$0.0.00')}</p>
//         {/*{expensesTotal > 0 && <p>total amount : {numeral(expensesTotal).format('$0.0.00')}</p>}*/}
//     </div>
// );

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount : visibleExpenses.length,
        expensesTotal : selectExpensesTotal(visibleExpenses)
        // expensesTotal : selectExpensesTotal(selectExpenses(state.expenses, state.filters))
        // expenses : selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);