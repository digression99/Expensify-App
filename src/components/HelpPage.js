import React from 'react';

const HelpPage = () => {
    return (
        <div>
            This is my help expense component.
        </div>
    );
};

export default HelpPage;