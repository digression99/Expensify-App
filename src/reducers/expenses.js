
// start by none of the expenses.
const expensesReducerDefaultState = [];

// reducer
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE' :
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            //return state.filter(expense => action.id !== expense.id);
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates // override the properties.
                    };
                } else {
                    return expense;
                }
            });
        default :
            return state;
    }
};