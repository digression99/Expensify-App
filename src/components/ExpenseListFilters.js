import React from 'react';
import { connect } from 'react-redux';
import {DateRangePicker} from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused : null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        // console.log(e.target.value);
        const sortBy = e.target.value;
        switch (sortBy) {
            case 'date':
                this.props.sortByDate();
                break;
            case 'amount':
                this.props.sortByAmount();
                break;
            default:
                console.log("error!");
                break;
        }
    };

    render() {
        // console.log('my props!', this.props);
        return (
            <div>
                <input type="text"
                       defaultValue={this.props.text}
                       onChange={this.onTextChange}
                />
                <label htmlFor="sort">SORT :
                    <select name="sort" id=""
                            onChange={this.onSortChange}
                    >
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                </label>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange} // function
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
}


// setup value and opChange for select.

const mapStateToProps = (state) => ({
    filters : state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter : (text) => dispatch(setTextFilter(text)),
    sortByDate : () => dispatch(sortByDate()),
    sortByAmount : () => dispatch(sortByAmount()),
    setStartDate : (startDate) => dispatch(setStartDate(startDate)),
    setEndDate : (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

// export default ExpenseListFilters;