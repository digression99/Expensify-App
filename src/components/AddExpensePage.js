import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        console.log('added!');
        // console.log(props.history);
        this.props.history.push('/'); // redirect.
    };
    render() {
        return (
            <div>
                <h1>Add expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>

        );
    }
}

// const AddExpensePage = (props) => {
//     return (
//         <div>
//             <h1>Add expense</h1>
//             <ExpenseForm
//                 onSubmit={(expense) => {
//                     // // props.dispatch(addExpense(expense));
//                     // props.onSubmit(expense);
//                     // console.log('added!');
//                     // // console.log(props.history);
//                     // props.history.push('/'); // redirect.
//                 }}
//             />
//         </div>
//     );
// };

const mapDispatchToProps = (dispatch) => ({addExpense : (expense) => dispatch(addExpense(expense))});

export default connect(undefined ,mapDispatchToProps)(AddExpensePage);