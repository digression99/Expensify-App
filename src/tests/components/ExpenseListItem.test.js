import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

// create test file
// grap imports
// render expensellistitem with one of the fixture items.
// create snapshot.

test('should render ExpenseListItem correctly', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').find('p').text()).toBe(`amount : ${expenses[0].amount}, createdAt : ${expenses[0].createdAt}`);
});