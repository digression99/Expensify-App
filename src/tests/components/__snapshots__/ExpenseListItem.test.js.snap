// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    amount : 
    $2.0
    , createdAt : 
    January 1st, 1970
  </p>
</div>
`;
